test/common/network/happy_eyeballs_connection_impl_test.cc

```diff
@@ -0,0 +1,1015 @@
+  EXPECT_CALL(*next_connections_.back(), setBufferLimits(42));
+  timeoutAndStartNextAttempt();
+
+  connectSecondAttempt();
+
+  // Verify that removeConnectionCallbacks calls are delegated to the remaining connection.
+  EXPECT_CALL(*created_connections_[1], setBufferLimits(420));
+  impl_->setBufferLimits(420);
+}
+
+TEST_F(HappyEyeballsConnectionImplTest, WriteBeforeLimit) {
+  startConnect();
+
+  Buffer::OwnedImpl data("hello world");
+  size_t length = data.length();
+  bool end_stream = false;
+
+  impl_->write(data, end_stream);
+  EXPECT_FALSE(impl_->aboveHighWatermark());
+
+  EXPECT_CALL(*created_connections_[0], setBufferLimits(length - 1));
+  impl_->setBufferLimits(length - 1);
+  EXPECT_TRUE(impl_->aboveHighWatermark());
+
+  // The call to write() will be replayed on the underlying connection.
+  EXPECT_CALL(*created_connections_[0], write(_, _))
+      .WillOnce(Invoke([](Buffer::Instance& data, bool end_stream) -> void {
+        EXPECT_EQ("hello world", data.toString());
+        EXPECT_FALSE(end_stream);
+      }));
+  connectFirstAttempt();
+}
+
+TEST_F(HappyEyeballsConnectionImplTest, WriteBeforeConnectOverLimit) {
+  startConnect();
+
+  Buffer::OwnedImpl data("hello world");
+  bool end_stream = false;
+
+  EXPECT_CALL(*created_connections_[0], setBufferLimits(data.length() - 1));
+  impl_->setBufferLimits(data.length() - 1);
+
+  impl_->write(data, end_stream);
+
+  // The call to write() will be replayed on the underlying connection.
+  EXPECT_CALL(*created_connections_[0], write(_, _))
+      .WillOnce(Invoke([](Buffer::Instance& data, bool end_stream) -> void {
+        EXPECT_EQ("hello world", data.toString());
+        EXPECT_FALSE(end_stream);
+      }));
+  connectFirstAttempt();
+}
+
+TEST_F(HappyEyeballsConnectionImplTest, AboveHighWatermark) {
+  startConnect();
+
+  EXPECT_FALSE(impl_->aboveHighWatermark());
+
+  connectFirstAttempt();
+
+  // Delegates to the connection once connected.
+  EXPECT_CALL(*created_connections_[0], aboveHighWatermark()).WillOnce(Return(true));
+  EXPECT_TRUE(impl_->aboveHighWatermark());
+}
+
+TEST_F(HappyEyeballsConnectionImplTest, SetConnectionStats) {
+  StrictMock<Stats::MockCounter> rx_total;
+  StrictMock<Stats::MockGauge> rx_current;
+  StrictMock<Stats::MockCounter> tx_total;
+  StrictMock<Stats::MockGauge> tx_current;
+  StrictMock<Stats::MockCounter> bind_errors;
+  StrictMock<Stats::MockCounter> delayed_close_timeouts;
+
+  Connection::ConnectionStats cs = {rx_total,   rx_current,   tx_total,
+                                    tx_current, &bind_errors, &delayed_close_timeouts};
+  EXPECT_CALL(*created_connections_[0], setConnectionStats(_))
+      .WillOnce(Invoke([&](const Connection::ConnectionStats& s) -> void { EXPECT_EQ(&s, &cs); }));
+  impl_->setConnectionStats(cs);
+
+  startConnect();
+
+  next_connections_.push_back(std::make_unique<StrictMock<MockClientConnection>>());
+  // setConnectionStats() should be applied to the newly created connection.
```

Comment:
hm, do we think that's correct?  I'd think we'd want to track bytes for SSL for abandoned connections not just the one we go wtih.